server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,745 @@
+     * @opensearch.internal
+     */
+    public static class Builder extends FieldMapper.Builder<Builder> {
+
+        private final IndexAnalyzers indexAnalyzers;
+
+        Builder(String name, IndexAnalyzers indexAnalyzers) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+            this.indexAnalyzers = indexAnalyzers;
+        }
+
+        public Builder(String name) {
+            this(name, null);
+        }
+
+        private FlatObjectFieldType buildFlatObjectFieldType(BuilderContext context, FieldType fieldType) {
+            NamedAnalyzer normalizer = Lucene.KEYWORD_ANALYZER;
+            return new FlatObjectFieldType(buildFullName(context), fieldType);
+        }
+
+        /**
+         * ValueFieldMapper is the subfield type for values in the Json.
+         * use a keywordfieldtype
+         */
+        private ValueFieldMapper buildValueFieldMapper(BuilderContext context, FieldType fieldType, FlatObjectFieldType fft) {
+            String fullName = buildFullName(context);
+            FieldType vft = new FieldType(fieldType);
+            KeywordFieldMapper.KeywordFieldType valueFieldType = new KeywordFieldMapper.KeywordFieldType(fullName + VALUE_SUFFIX, vft);
+            fft.setValueFieldType(valueFieldType);
+            return new ValueFieldMapper(vft, valueFieldType);
+        }
+
+        /**
+         * ValueAndPathFieldMapper is the subfield type for path=value format in the Json.
+         * also use a keywordfieldtype
+         */
+        private ValueAndPathFieldMapper buildValueAndPathFieldMapper(BuilderContext context, FieldType fieldType, FlatObjectFieldType fft) {
+            String fullName = buildFullName(context);
+            FieldType vft = new FieldType(fieldType);
+            KeywordFieldMapper.KeywordFieldType ValueAndPathFieldType = new KeywordFieldMapper.KeywordFieldType(
+                fullName + VALUE_AND_PATH_SUFFIX,
+                vft
+            );
+            fft.setValueAndPathFieldType(ValueAndPathFieldType);
+            return new ValueAndPathFieldMapper(vft, ValueAndPathFieldType);
+        }
+
+        @Override
+        public FlatObjectFieldMapper build(BuilderContext context) {
+            FieldType fieldtype = new FieldType(Defaults.FIELD_TYPE);
+            FlatObjectFieldType fft = buildFlatObjectFieldType(context, fieldtype);
+            return new FlatObjectFieldMapper(
+                name,
+                Defaults.FIELD_TYPE,
+                fft,
+                buildValueFieldMapper(context, fieldtype, fft),
+                buildValueAndPathFieldMapper(context, fieldtype, fft),
+                MultiFields.empty(),
+                CopyTo.empty(),
+                this
+            );
+        }
+    }
+
+    public static final TypeParser PARSER = new TypeParser((n, c) -> new Builder(n, c.getIndexAnalyzers()));
+
+    /**
+     * Creates a new TypeParser for flatObjectFieldMapper that does not use ParameterizedFieldMapper
```

Comment:
Why?

Reply:
The FlatObject FieldMapper is extended from DynamicKeyFieldMapper, which is extended from FieldMapper. The FieldMapper class does not have a TypeParser class is likely due to differences in the requirements and functionality of these classes. 

FieldMapper is responsible for mapping simple fields without any parameters, such as string or integer fields. In this case, there may not be a need for a TypeParser class because the field types are already well-defined and do not require any additional parsing or validation.

ParameterizedFieldMapper is a subclass of FieldMapper and is responsible for mapping fields that have parameters, such as the geo_point field type which requires latitude and longitude values. The TypeParser class in ParameterizedFieldMapper is responsible for parsing and validating these parameters.

In FlatObject FieldMapper, we are intended to open parameters in the future, this is why we will need a TypeParser class in this first PR.